### PR TITLE
Fix image data to correctly get images from articles

### LINKED ISSUE #89: no attribute-value-not-found error for wikibase-term IMAGE-DATA
Currently:

    $ telnet nauru 8023
    (get "wikibase-term" "The Tiger (1978 film)" (:CODE "IMAGE-DATA"))
    ((:error attribute-value-not-found :reply "Unknown"))

Every wikibase-term has an IMAGE-DATA attribute with a valid value.  When an article doesn't have any images (as with `The Tiger (1978 film)`), the correct return value is the empty list, `()`. IMAGE-DATA should not throw `attribute-value-not-found`.

### LINKED ISSUE #110: IMAGE-DATA wrongly returning no value in some cases
Why no image data?  There are lots images in the https://en.wikipedia.org/wiki/Flower article.

    (get "wikibase-term" "Flower" (:CODE "IMAGE-DATA"))
    ((:error attribute-value-not-found :message "Unknown"))

### PR DESCRIPTION
**Pull request by sherryy**
Fix and test image-data attributes to return a prioritized list of images or an empty list if there is no images at all

Fixes #110 and fixes #89

### CODE REVIEW

**Comment by chsamlee** [comment]
Looks good to me.

File: tests/test_resolvers.py [hunk 1]
```diff
@@ -103,6 +103,31 @@ def test_error_resolver(self):
         err = self.kb.get('wikibase-person', 'Barack Obama', 'death-date')
         self.assertEqual(str(err), alive_err)
 
+    def test_image_data_best_infobox_no_caption(self):
+        article = u"Bill Clinton"
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: instead of defining unicode literals via `u"String"`, it might be a good idea to import unicode_literals at the top of the file.

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_resolvers
----------------------------------

Tests for `resolvers` module.
"""

from __future__ import unicode_literals  # <---

try:
    import unittest2 as unittest
except ImportError:
    import unittest
```

This feature makes every literal string a Unicode string. For `test_resolvers.py` this would be very useful.

File: tests/test_resolvers.py [hunk 2]
```diff
@@ -103,6 +103,31 @@ def test_error_resolver(self):
...
+        attr = u"image-data"
+        best_img = self.term_resolver.image(article, attr)[0].val
+        self.assertEqual(str(best_img["file"]), "44_Bill_Clinton_3x4.jpg")
+
+    def test_image_data_best_article_caption(self):
+        article = u"Flower"
+        attr = u"image-data"
+        best_img = self.term_resolver.image(article, attr)[0].val
+        self.assertEqual(str(best_img["file"]), "Flower_poster_2.jpg")
+        self.assertEqual(str(best_img["caption"])[:-2], "A poster with flowers or clusters of flowers produced by twelve species of flowering plants from different families")
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: line is too long; split to keep ~80 chars

File: tests/test_resolvers.py [hunk 3]
```diff
@@ -103,6 +103,31 @@ def test_error_resolver(self):
...
+        attr = u"image-data"
+        best_img = self.term_resolver.image(article, attr)[0].val
+        self.assertEqual(str(best_img["file"]), "Wikipedia-logo-v2.svg")
+
+    def test_image_data_image_flag(self):
+        article = u"France"
+        attr = u"image-data"
+        best_img = self.term_resolver.image(article, attr)[0].val
+        self.assertEqual(str(best_img["file"]), "Flag_of_France.svg")
+
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: You should add a test to check that image filenames are escaped using url escaping. You can use the Yuri Manin example:

```
(get "wikibase-term" "Yuri I. Manin" (:CODE "IMAGE-DATA"))
```

Comma is escaped: `:file "Juri_Manin%2C_Ksenia_Semenova.jpeg"`

File: wikipediabase/article.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 from wikipediabase.fetcher import get_fetcher
 from wikipediabase.log import Logging
 from wikipediabase.util import Expiry, LRUCache, fromstring, tostring
+import re
 
 
 # XXX: also support images.
```
Comment (alvaromorales) [inline wikipediabase/article.py]: I think you can now remove this comment :-)

File: wikipediabase/infobox.py [hunk 5]
```diff
@@ -56,6 +56,11 @@ def get_from_markup(self, attr):
             if k == attr:
                 return v
 
+    def get_infobox_image(self):
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: Can an infobox have multiple images? If so, this method should be called `get_infobox_images` and should return a list

File: wikipediabase/resolvers/term.py [hunk 6]
```diff
@@ -47,25 +48,35 @@ def coordinates(self, article, _):
     @provide(name="image-data")
     def image(self, article, attribute):
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: rename this method to `image_data` for consistency; a future change will implement `image_live`

File: wikipediabase/resolvers/term.py [hunk 7]
```diff
@@ -47,25 +48,35 @@ def coordinates(self, article, _):
...
+        for article_img in article_imgs:
+            if "File:" in article_img:
+                fnam = article_img.split("File:")[1]
+            caption = re.findall(r"\|.*\]\]", fnam)
+            fnam = re.findall(r".*\.(?:ogv|png|jpg|jpeg|gif|tif|svg)", fnam)[0].replace(" ", "_")
+            if caption:
+                cap = caption[0].split("|")[-1]
+                cap = cap if "alt =" not in cap else ""
+            imgs.append(lispify({"caption": markup_unlink(cap), "file": urllib.quote(fnam.encode('utf8'))}))
+        return imgs
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: There's a bug in the return value of image-data, we have an extra level of nesting:

```python
In [2]: wikipediabase('(get "wikibase-term" "Yuri I. Manin" (:CODE "IMAGE-DATA"))')
Out[2]: u'(((:caption "Yuri Manin with his wife Ksenia Semenova at the International Congress of Mathematicians|ICM 2006 in Madrid" :file "Juri_Manin%2C_Ksenia_Semenova.jpeg")))'
```

[we should return a list of plists, but we're returning a list of lists of plists]

You should add a test that checks that the output of `wikipediabase(query)` is in the correct format.

For starters, instead of appending lispified dicts to a Python list and then returning a Python list, you should just `imgs.append(lispify(image_data))` (e.g. line 67) and `return lispify(imgs)`. This makes unit testing a little harder because you can't index into a LispList, but you can `self.assertIn`.

But the real issue is not in `image-data`. See `wikipediabase/knowledgebase.py` L29. The _correct_ way to solve this is to standardize the return values of resolvers and only call lispify in frontend. We should do this in a future enhancement, though. This makes me sad, but we might need a temporary nasty hack:

```diff
diff --git a/wikipediabase/knowledgebase.py b/wikipediabase/knowledgebase.py
index f117c76..2f7f718 100644
--- a/wikipediabase/knowledgebase.py
+++ b/wikipediabase/knowledgebase.py
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-

 from wikipediabase.classifiers import classify
-from wikipediabase.lispify import lispify
+from wikipediabase.lispify import lispify, LispList
 from wikipediabase.provider import Provider, provide
 from wikipediabase.resolvers import resolve, resolver_attributes
 from wikipediabase.sort_symbols import sort_by_length, sort_named
@@ -26,7 +26,12 @@ class KnowledgeBase(Provider):
         :param attr: the attribute to get
         :returns: the attribute's value or an error, lispified
         """
-        return lispify([resolve(cls, symbol, attr)])
+        # TODO: fix by standardizing the return value of resolvers
+        # lispify should only be called in frontend
+        result = resolve(cls, symbol, attr)
+        if isinstance(result, LispList):
+            return result
+        return lispify([result])

     @provide(name="get-attributes")
     def get_attributes(self, cls, symbol):
```

```diff
diff --git a/wikipediabase/resolvers/term.py b/wikipediabase/resolvers/term.py
index 4fddf21..1b56842 100644
--- a/wikipediabase/resolvers/term.py
+++ b/wikipediabase/resolvers/term.py
@@ -43,7 +43,7 @@ class TermResolver(BaseResolver):
             nlat = self._dton(totext(lat))
             nlon = self._dton(totext(lon))

-            return lispify([nlat, nlon], typecode='coordinates')
+            return lispify([lispify([nlat, nlon], typecode='coordinates')])

     @provide(name="image-data")
     def image(self, article, attribute):
@@ -64,7 +64,7 @@ class TermResolver(BaseResolver):
                 image_data = {"file": urllib.quote(fnam)}
                 if caption:
                     image_data["caption"] = caption
-                imgs.append(lispify(image_data))
+                imgs.append(image_data)

         article_imgs = get_article(article).get_article_images()
         for article_img in article_imgs:
@@ -75,8 +75,8 @@ class TermResolver(BaseResolver):
             if caption:
                 cap = caption[0].split("|")[-1]
                 cap = cap if "alt =" not in cap else ""
-            imgs.append(lispify({"caption": markup_unlink(cap), "file": urllib.quote(fnam.encode('utf8'))}))
-        return imgs
+            imgs.append({"caption": markup_unlink(cap), "file": urllib.quote(fnam.encode('utf8'))})
+        return lispify(imgs)

     @provide(name='number')
     def number(self, article, _):
@@ -147,7 +147,7 @@ class TermResolver(BaseResolver):
     def word_count(self, article, attribute):
         self.log().info("Trying 'word-count' tag from static resolver.")
         self._tag = "html"
-        return len(self._words(get_fetcher().html_source(article)))
+        return lispify([[len(self._words(get_fetcher().html_source(article)))]])

     @staticmethod
     def _dton(s):
```

```diff
diff --git a/tests/test_resolvers.py b/tests/test_resolvers.py
index f4cd79e..cc1b064 100644
--- a/tests/test_resolvers.py
+++ b/tests/test_resolvers.py
@@ -8,6 +8,8 @@ test_resolvers
 Tests for `resolvers` module.
 """

+from __future__ import unicode_literals
+
 try:
     import unittest2 as unittest
 except ImportError:
@@ -21,7 +23,7 @@ from wikipediabase.frontend import Frontend
 from wikipediabase.knowledgebase import KnowledgeBase
 from wikipediabase.resolvers.person import first_paren

-ARTICLE_BODY = u"""A ninja (忍者?) or shinobi (忍び?) was a covert agent or mercenary
+ARTICLE_BODY = """A ninja (忍者?) or shinobi (忍び?) was a covert agent or mercenary
 in feudal Japan. The functions of the ninja included espionage,
 sabotage, infiltration, and assassination, and open combat in certain
 situations.[1] Their covert methods of waging war contrasted the ninja
@@ -104,29 +106,21 @@ class TestResolvers(unittest.TestCase):
         self.assertEqual(str(err), alive_err)

     def test_image_data_best_infobox_no_caption(self):
-        article = u"Bill Clinton"
-        attr = u"image-data"
-        best_img = self.term_resolver.image(article, attr)[0].val
-        self.assertEqual(str(best_img["file"]), "44_Bill_Clinton_3x4.jpg")
+        imgs = self.term_resolver.image('Bill Clinton', 'image-data')
+        self.assertIn("44_Bill_Clinton_3x4.jpg", imgs)

     def test_image_data_best_article_caption(self):
-        article = u"Flower"
-        attr = u"image-data"
-        best_img = self.term_resolver.image(article, attr)[0].val
-        self.assertEqual(str(best_img["file"]), "Flower_poster_2.jpg")
-        self.assertEqual(str(best_img["caption"])[:-2], "A poster with flowers or clusters of flowers produced by twelve species of flowering plants from different families")
+        imgs = self.term_resolver.image('Flower', 'image-data')
+        self.assertIn("Flower_poster_2.jpg", imgs)
+        self.assertIn("A poster with flowers or clusters of flowers produced by twelve species of flowering plants from different families", imgs)

     def test_image_data_wiki_logo_in_article(self):
-        article = u"Wikipedia logo"
-        attr = u"image-data"
-        best_img = self.term_resolver.image(article, attr)[0].val
-        self.assertEqual(str(best_img["file"]), "Wikipedia-logo-v2.svg")
+        imgs = self.term_resolver.image('Wikipedia logo', 'image-data')
+        self.assertIn("Wikipedia-logo-v2.svg", imgs)

     def test_image_data_image_flag(self):
-        article = u"France"
-        attr = u"image-data"
-        best_img = self.term_resolver.image(article, attr)[0].val
-        self.assertEqual(str(best_img["file"]), "Flag_of_France.svg")
+        imgs = self.term_resolver.image('France', 'image-data')
+        self.assertIn("Flag_of_France.svg", imgs)

     def _ans_match(self, lst, just=None):
         """
```

Again, this is only a temporary fix so we can push the changes to IMAGE-DATA.

**Comment by alvaromorales** [comment]
This is looking good! I left some inline comments.

**Comment by alvaromorales** [comment]
Using the 'Random article' button, I found interesting cases (I also fixed some merge conflicts in the process):

* Missing image

    ```
    (get "wikibase-term" "Sector 2 (Bucharest)" (:code "IMAGE-DATA"))
    ```
    https://en.wikipedia.org/wiki/Sector_2_(Bucharest)

* Markup (illegal characters) in caption

    ```
    (get "wikibase-term" "Sinking of the Rainbow Warrior" (:code "IMAGE-DATA"))
    ((:caption "A 2D profile drawing of \'\'Rainbow Warrior (1978)|Rainbow Warrior\'\'." [...]
    ```

* Missing captions

    ```
    (get "wikibase-term" "Microblepsis violacea" (:code "IMAGE-DATA"))
    ((:file "Microblepsis_violacea.jpg") (:file "Microblepsis_violacea1.jpg"))
    ```
    https://en.wikipedia.org/wiki/Microblepsis_violacea

    ```
    (get "wikibase-term" "Dusky flying fox" (:code "IMAGE-DATA"))
    ((:file "Pteropus.jpg"))
    ```
    https://en.wikipedia.org/wiki/Microblepsis_violacea

* Runtime error

    `UnboundLocalError: local variable 'cap' referenced before assignment`

    ```
    (get "wikibase-term" "Coast Range (EPA ecoregion)" (:code "IMAGE-DATA"))
    ```

* Images in markup with `Image:` syntax

    ```
    (get "wikibase-term" "Lymphoepithelial lesion" (:code "IMAGE-DATA"))
    ```
    Also see [Spigelia marilandica](https://en.wikipedia.org/wiki/Spigelia_marilandica)

* Image extension regex should be case insensitive

    See https://en.wikipedia.org/wiki/Sagar_Kashyap (`| image = Sagar Kashyap.JPG`)

### Harder bugs

* Map missing because of nested image attribute for https://en.wikipedia.org/wiki/Tur_Langton

    ```
    |static_image = |static_image_caption= 
    }}
    
    [[File:Tur Langton - geograph.org.uk - 150667.jpg|thumbnail|A street in Tur Langton.]]
    '''Tur Langton''' (derived from the Anglo-Saxon word for an enclosure, meaning "long town")<ref>{{cite    episode
    ```

* Image defined in attributes like `cover` or `logo`

     See https://en.wikipedia.org/wiki/Hardware_(album) and https://en.wikipedia.org/wiki/Montana_Department_of_Fish,_Wildlife_and_Parks

    @sfelshin do we consider this a valid image for IMAGE-DATA? If so, we might need to stop trying to find `image` in the attribute name, and only look for values with an image file extension.

**Comment by alvaromorales** [comment]
Almost there!

* Missing images
    * Dinner train
    * NRP Adamastor
    * Shahid Afridi Foundation  --> logo
    * Thoughts of Love --> cover
    * Poffer --> "best" (first) image missing
* Wrong caption
    * WOIO --> "180px", not a caption
* Missing caption
    * Susan Pamerleau
* Repeated image, wrong caption
    * Wisconsin gubernatorial election, 2014
    * Lady Jewelpet
* Runtime errors
    * Cingulum (brain) --> `IndexError: list index out of range`

### Edge cases

Everything In Wikipedia is inconsistent. I don't think these are worth fixing:

* Missing caption
    * `cap` instead of `caption`: Pasco County, Florida

**Comment by alvaromorales** [comment]
Also, can you add a unit test for #89?

**Comment by alvaromorales** [comment]
pep8 style check failed, run `invoke lint` locally to fix.

File: wikipediabase/article.py [hunk 8]
```diff
@@ -87,6 +85,13 @@ def first_paragraph(self, keep_html=False):
 
         return None
 
+    def get_article_images(self):
+        markup = self.markup_source()
+        img_pattern = re.compile("^\[\[(?:File|Image):.*\.(?:ogv|png|jpg|jpeg|gif|tif|svg).*\]\]",
+                                 re.IGNORECASE | re.MULTILINE)
```
Comment (alvaromorales) [inline wikipediabase/article.py]: always use named keyword arguments in case someone adds another argument later; change to `flags=re.IGNORECASE ...`

File: wikipediabase/infobox.py [hunk 9]
```diff
@@ -66,6 +66,21 @@ def get_from_markup(self, attr):
             if k == attr:
                 return v
 
+    def get_infobox_captions(self):
+        infobox_captions = []
+        for k, v in self._markup_items():
+            if ("map" not in k and (v != "") and re.search(r"caption\d*", k)) or ("image" in k and "caption" in k):
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: You should add a comment here explaining the logic to find captions.

`v != ""` won't always check if an attribute has a value. Use the `has_value` method instead.

File: wikipediabase/infobox.py [hunk 10]
```diff
@@ -66,6 +66,21 @@ def get_from_markup(self, attr):
...
+                infobox_captions.append(v)
+        return infobox_captions
+
+    def get_infobox_images(self):
+        infobox_imgs = []
+        image_extention = re.compile("\.(?:jpe?g|pdf|gif|png|bpg|svg)", re.IGNORECASE)
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: typo: `image_extension`

File: wikipediabase/infobox.py [hunk 11]
```diff
@@ -66,6 +66,21 @@ def get_from_markup(self, attr):
...
+        for k, v in self._markup_items():
+            if ("image" in k or "logo" in k or "cover" in k) and re.search(image_extention, v) and ("map" not in k):
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: You should probably put `"image"`, `"logo"`, etc in some variable called `IMAGE_PATTERNS = [ ...]`, and then check in for loop. This will make it easier to add patterns later on as we discover more cases.

File: wikipediabase/resolvers/term.py [hunk 12]
```diff
@@ -1,5 +1,6 @@
-# -*- coding: utf-8 -*-
+# -*- coding:utf-8 -*-
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: ^ undo this change; the space should be there

File: wikipediabase/resolvers/term.py [hunk 13]
```diff
@@ -42,30 +43,52 @@ def coordinates(self, article, _, **options):
...
+                    fnam = re.findall(re.compile(".*\.(?:ogv|png|jpg|jpeg|gif|tif|svg)",
+                                                 re.IGNORECASE), fnam)[0]
+
+                    if "File:" in img:
+                        fnam = fnam.split("File:")[1]
+                    if "{{!}}border" in img:
+                        fnam = fnam.split("{{!}}border")[0]
+                    caption = []
+                    if info_captions:
+                        if info_imgs.index(img) <= (len(info_captions) - 1):
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: this needs a comment explaining how you align images to captions

File: wikipediabase/resolvers/term.py [hunk 14]
```diff
@@ -42,30 +43,52 @@ def coordinates(self, article, _, **options):
...
+                    if caption:
+                        image_data["caption"] = markup_unlink(caption)
+                    imgs.append(image_data)
+        article_imgs = get_article(article).get_article_images()
+        for article_img in article_imgs:
+            fnam = article_img
+            if "File:" in fnam:
+                fnam = fnam.split("File:")[1]
+            if "Image:" in fnam:
+                fnam = fnam.split("Image:")[1]
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: repeated logic from above; move to helper method

**Comment by alvaromorales** [comment]
Great work! I left some inline comments.

**Comment by alvaromorales** [comment]
Minor nitpicks. After that, this is good to go!

File: wikipediabase/infobox.py [hunk 15]
```diff
@@ -67,6 +67,25 @@ def get_from_markup(self, attr):
...
+    def get_infobox_captions(self):
+        '''check either attibute's name contain both 'image' and 'caption' or contains caption but not map'''
+        infobox_captions = []
+        for k, v in self._markup_items():
+            if ("map" not in k and self.has_value(k) and re.search(r"caption\d*", k)) or ("image" in k and "caption" in k):
+                infobox_captions.append(v)
+        return infobox_captions
+
+    def get_infobox_images(self):
+        '''attribute contains 'image','logo', or 'cover' and value contains valid file extension, excluding map case'''
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: Please use three double-quotes (`"""`) for docstrings everywhere in this PR

File: wikipediabase/resolvers/term.py [hunk 16]
```diff
@@ -51,28 +52,52 @@ def coordinates(self, article, _, **options):
...
+            caption = re.findall(r"\|.*\]\]", article_img)
+            fnam = self.extract_image_url(article_img)
+            image_data = {"file": urllib.quote(fnam.encode('utf8'))}
+            if caption:
+                caption = caption[0].split("|")[-1][:-2]
+            imgs.append(self.combine_image_caption(fnam, caption))
+        return imgs
+
+    def extract_image_url(self, img):
+        """fomat image url"""
```
Comment (alvaromorales) [inline wikipediabase/resolvers/term.py]: typo, `fomat` --> `format`

**Comment by alvaromorales** [comment]
Some comments from the previous review weren't addressed

**Comment by alvaromorales** [comment]
Looks great :rocket: 

Congrats on the great work; this change was a tough one with many edge cases!
